gateway/src/apicast/policy_loader.lua

```diff
@@ -0,0 +1,199 @@
+local error = error
+local type = type
+local loadfile = loadfile
+local getenv = os.getenv
+local insert = table.insert
+local setmetatable = setmetatable
+local concat = table.concat
+
+local _M = { }
+
+local searchpath = package.searchpath
+local root_loaded = package.loaded
+
+local root_require = require
+local pcall = pcall
+
+local preload = package.preload
+
+--- create a require function not using the global namespace
+-- loading code from policy namespace should have no effect on the global namespace
+-- but poliocy can load shared libraries that would be cached globally
+local function gen_require(package)
+
+  local function not_found(modname, err)
+    return error(format("module '%s' not found:%s", modname, err), 0)
+  end
+
+  --- helper function to safely use the native require function
+  local function fallback(modname)
+    local mod
+
+    mod = package.loaded[modname]
+
+    if not mod then
+      ngx.log(ngx.DEBUG, 'native require for: ', modname)
+      mod = root_require(modname)
+    end
+
+    return mod
+  end
+
+  --- helper function to find and return correct loader for a module
+  local function find_loader(modname)
+    local loader, file, err, ret
+
+    -- http://www.lua.org/manual/5.2/manual.html#pdf-package.searchers
+
+    -- When looking for a module, require calls each of these searchers in ascending order,
+    for i=1, #package.searchers do
+      -- with the module name (the argument given to require) as its sole parameter.
+      ret, err = package.searchers[i](modname)
+
+      -- The function can return another function (the module loader)
+      -- plus an extra value that will be passed to that loader,
+      if type(ret) == 'function' then
+        loader = ret
+        file = err
+        break
+      -- or a string explaining why it did not find that module
+      elseif type(ret) == 'string' then
+        err = ret
+      end
+      -- (or nil if it has nothing to say).
+    end
+
+    return loader, file, err
+  end
+
+  --- reimplemented require function
+  -- - return a module if it was already loaded (globally or locally)
+  -- - try to find loader function
+  -- - fallback to global require
+  -- @tparam string modname module name
+  -- @tparam boolean exclusive load only policy code, turns off the fallback loader
+  return function(modname, exclusive)
+    ngx.log(ngx.DEBUG, 'sandbox require: ', modname)
+    local mod = root_loaded[modname]
+
+    if mod then return mod end
+
+    local loader, file, err = find_loader(modname)
+
+    if loader then
+      ngx.log(ngx.DEBUG, 'sandboxed require for: ', modname, ' file: ', file)
+      mod = loader(modname, file)
+    elseif not exclusive then
+      ngx.log(ngx.DEBUG, 'fallback loader for: ', modname, ' error: ', err)
+      mod = fallback(modname)
+    else
+      return not_found(modname, err)
+    end
+
+    if mod ~= nil then
+      package.loaded[modname] = mod
+    else
+      package.loaded[modname] = true
```

Comment:
Why this assignment?

Reply:
Will add there the comment from the docs 👍